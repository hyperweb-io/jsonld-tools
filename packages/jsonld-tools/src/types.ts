/**
 * Types for the JSON-LD Builder API
 *
 * Provides comprehensive type definitions for the fluent builder interface
 * that maintains compatibility with existing filtering logic while adding
 * new configuration-first approach.
 */

// Import and re-export core types from local jsonld utils
import type { JsonLdEntity, JsonLdGraph, PropertyFilterConfig, PropertyFilterRule } from './jsonld-utils';

export type { JsonLdEntity, JsonLdGraph, PropertyFilterConfig, PropertyFilterRule };

/**
 * Filter function type for JSON-LD entities
 */
export type JsonLdFilter = (entity: JsonLdEntity) => boolean;

/**
 * Updated filter options with renamed properties for clarity
 */
export interface JsonLdFilterOptions {
	/** Include only these entity types */
	includeTypes?: string[];

	/** Exclude these entity types */
	excludeTypes?: string[];

	/** Include only entities with these IDs */
	includeIds?: string[];

	/** Exclude entities with these IDs */
	excludeIds?: string[];

	/** Custom filter function */
	customFilter?: JsonLdFilter;

	/** Maximum number of entities to include */
	maxEntities?: number;

	/** Include only entities with specific properties */
	requiredProperties?: string[];

	/** Exclude entities with specific properties (renamed from excludeProperties) */
	excludeEntitiesWithProperties?: string[];
}

/**
 * Configuration for property filtering by entity IDs
 */
export interface PropertyFilterByIdRule {
	/** Entity IDs to apply this rule to */
	entityIds: string[];
	/** Properties to include (if specified, only these are included) */
	include?: string[];
	/** Properties to exclude */
	exclude?: string[];
}

/**
 * Configuration for property filtering by entity types
 */
export interface PropertyFilterByTypeRule {
	/** Entity types to apply this rule to */
	entityTypes: string[];
	/** Properties to include (if specified, only these are included) */
	include?: string[];
	/** Properties to exclude */
	exclude?: string[];
}

export interface PopulateProperty {
	property: string;
	entities: JsonLdEntity[];
}

export interface PopulateConfig {
	[entityId: string]: PopulateProperty[];
}

/**
 * Immutable configuration state for JSON-LD processing
 */
export interface JsonLdConfig {
	/** Base graph to process */
	baseGraph?: JsonLdGraph;

	/** Entity filtering options */
	filters?: JsonLdFilterOptions;

	/** Root IDs for subgraph extraction */
	subgraphRoots?: string[];

	/** Property-level filtering configuration */
	propertyFilters?: PropertyFilterConfig;

	/** Property filtering rules by entity IDs */
	propertyFiltersByIds?: PropertyFilterByIdRule[];

	/** Property filtering rules by entity types */
	propertyFiltersByTypes?: PropertyFilterByTypeRule[];

	/** Additional entities to include */
	additionalEntities?: JsonLdEntity[];

	/** Custom transformation functions */
	pipes?: PipeFunction[];

	/** Populate configuration (contains data to populate) */
	populateConfig?: PopulateConfig;
}

/**
 * Options for the build() method
 */
export interface BuildOptions {
	/** Whether to pretty-print the JSON output */
	prettyPrint?: boolean;

	/** Custom context URL (defaults to 'https://schema.org') */
	contextUrl?: string;

	/** Whether to wrap output in a script tag */
	withScriptTag?: boolean;

	/** ID for the script tag (only used if withScriptTag is true) */
	scriptId?: string;
}

/**
 * Custom transformation function type
 */
export type PipeFunction = (graph: JsonLdGraph) => JsonLdGraph;

/**
 * Configuration builder interface (immutable operations)
 */
export interface IJsonLdConfigBuilder {
	/** Set the base graph to process */
	baseGraph(graph: JsonLdGraph): this;

	/** Include only entities with these IDs */
	includeIds(ids: string[]): this;

	/** Exclude entities with these IDs */
	excludeIds(ids: string[]): this;

	/** Include only these entity types */
	includeTypes(types: string[]): this;

	/** Exclude these entity types */
	excludeTypes(types: string[]): this;

	/** Apply custom filter function */
	customFilter(fn: JsonLdFilter): this;

	/** Limit maximum number of entities */
	maxEntities(max: number): this;

	/** Include only entities with these properties */
	requiredProperties(props: string[]): this;

	/** Exclude entities with these properties */
	excludeEntitiesWithProperties(props: string[]): this;

	/** Extract subgraph starting from these root IDs */
	subgraph(rootIds: string[]): this;

	/** Filter properties for specific entity IDs */
	filterPropertiesByIds(entityIds: string[], rule: { include?: string[]; exclude?: string[] }): this;

	/** Filter properties for specific entity types */
	filterPropertiesByTypes(entityTypes: string[], rule: { include?: string[]; exclude?: string[] }): this;

	/** Add additional entities */
	addEntities(entities: JsonLdEntity[]): this;

	/** Add custom transformation pipe */
	pipe(fn: PipeFunction): this;

	/** Configure entity population */
	populateEntities(entityIds: string[], populateConfig: PopulateConfig): this;

	/** Clear both includeIds and excludeIds */
	clearIds(): this;

	/** Clear both includeTypes and excludeTypes */
	clearTypes(): this;

	/** Clear both requiredProperties and excludeEntitiesWithProperties */
	clearPropertyRequirements(): this;

	/** Clear both propertyFiltersByIds and propertyFiltersByTypes */
	clearPropertyFilters(): this;

	/** Clear subgraphRoots */
	clearSubgraph(): this;

	/** Clear entire configuration except baseGraph */
	clearAll(): this;

	/** Get the current configuration */
	getConfig(): JsonLdConfig;
}
