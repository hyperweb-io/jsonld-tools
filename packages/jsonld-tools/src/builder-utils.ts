/**
 * Utility functions for JSON-LD builder
 *
 * Provides configuration merging, validation, and transformation utilities
 * for the JSON-LD builder system.
 */

import type {
  JsonLdEntity,
  JsonLdGraph,
  PropertyFilterConfig,
  PropertyFilterRule,
} from './jsonld-utils';
import type {
  JsonLdConfig,
  JsonLdFilterOptions,
  PopulateConfig,
  PropertyFilterByIdRule,
  PropertyFilterByTypeRule,
} from './types';

/**
 * Merge two configurations, with the second taking precedence
 */
export function mergeConfigs(base: JsonLdConfig, override: JsonLdConfig): JsonLdConfig {
  return {
    // Base graph: override takes precedence
    baseGraph: override.baseGraph ?? base.baseGraph,

    // Filters: merge filter options
    filters: mergeFilterOptions(base.filters, override.filters),

    // Additional entities: combine arrays
    additionalEntities: combineArrays(base.additionalEntities, override.additionalEntities),

    // Pipes: combine arrays
    pipes: combineArrays(base.pipes, override.pipes),

    // Populate config: override takes precedence
    populateConfig: override.populateConfig ?? base.populateConfig,
  };
}

/**
 * Merge filter options, merging arrays and using override for single values
 */
function mergeFilterOptions(
  base?: JsonLdFilterOptions,
  override?: JsonLdFilterOptions
): JsonLdFilterOptions | undefined {
  if (!base && !override) return undefined;
  if (!base) return override;
  if (!override) return base;

  return {
    // Merge arrays instead of replacing
    includeTypes: combineArrays(base.includeTypes, override.includeTypes),
    excludeTypes: combineArrays(base.excludeTypes, override.excludeTypes),
    includeIds: combineArrays(base.includeIds, override.includeIds),
    excludeIds: combineArrays(base.excludeIds, override.excludeIds),
    requiredProperties: combineArrays(base.requiredProperties, override.requiredProperties),
    excludeEntitiesWithProperties: combineArrays(
      base.excludeEntitiesWithProperties,
      override.excludeEntitiesWithProperties
    ),
    subgraphRoots: combineArrays(base.subgraphRoots, override.subgraphRoots),
    propertyFilters: combinePropertyFilters(base.propertyFilters, override.propertyFilters),
    propertyFiltersByIds: combineArrays(base.propertyFiltersByIds, override.propertyFiltersByIds),
    propertyFiltersByTypes: combineArrays(
      base.propertyFiltersByTypes,
      override.propertyFiltersByTypes
    ),

    // Single values - override takes precedence
    customFilter: override.customFilter ?? base.customFilter,
    maxEntities: override.maxEntities ?? base.maxEntities,
  };
}

/**
 * Combine property filter configurations
 */
function combinePropertyFilters(
  base?: PropertyFilterConfig,
  override?: PropertyFilterConfig
): PropertyFilterConfig | undefined {
  if (!base && !override) return undefined;
  if (!base) return override;
  if (!override) return base;
  return [...base, ...override];
}

/**
 * Combine arrays, handling undefined values
 */
function combineArrays<T>(base?: T[], override?: T[]): T[] | undefined {
  if (!base && !override) return undefined;
  if (!base) return override;
  if (!override) return base;
  return [...base, ...override];
}

/**
 * Convert property filter by ID rules to PropertyFilterConfig
 */
export function convertPropertyFiltersByIds(rules: PropertyFilterByIdRule[]): PropertyFilterRule[] {
  return rules.flatMap((rule) =>
    rule.entityIds.map((entityId) => ({
      selector: { '@id': entityId },
      include: rule.include,
      exclude: rule.exclude,
    }))
  );
}

/**
 * Convert property filter by type rules to PropertyFilterConfig
 */
export function convertPropertyFiltersByTypes(
  rules: PropertyFilterByTypeRule[]
): PropertyFilterRule[] {
  return rules.flatMap((rule) =>
    rule.entityTypes.map((entityType) => ({
      selector: { '@type': entityType },
      include: rule.include,
      exclude: rule.exclude,
    }))
  );
}

/**
 * Build complete property filter configuration from all sources
 */
export function buildPropertyFilterConfig(config: JsonLdConfig): PropertyFilterConfig | undefined {
  const filters: PropertyFilterRule[] = [];

  // Add base property filters
  if (config.filters?.propertyFilters) {
    filters.push(...config.filters.propertyFilters);
  }

  // Add property filters by IDs
  if (config.filters?.propertyFiltersByIds) {
    filters.push(...convertPropertyFiltersByIds(config.filters.propertyFiltersByIds));
  }

  // Add property filters by types
  if (config.filters?.propertyFiltersByTypes) {
    filters.push(...convertPropertyFiltersByTypes(config.filters.propertyFiltersByTypes));
  }

  return filters.length > 0 ? filters : undefined;
}

/**
 * Validate configuration for common issues
 */
export function validateConfig(config: JsonLdConfig): string[] {
  const errors: string[] = [];

  // Check for conflicting include/exclude types
  if (config.filters?.includeTypes && config.filters?.excludeTypes) {
    const overlap = config.filters.includeTypes.filter((type) =>
      config.filters!.excludeTypes!.includes(type)
    );
    if (overlap.length > 0) {
      errors.push(`Conflicting include/exclude types: ${overlap.join(', ')}`);
    }
  }

  // Check for conflicting include/exclude IDs
  if (config.filters?.includeIds && config.filters?.excludeIds) {
    const overlap = config.filters.includeIds.filter((id) =>
      config.filters!.excludeIds!.includes(id)
    );
    if (overlap.length > 0) {
      errors.push(`Conflicting include/exclude IDs: ${overlap.join(', ')}`);
    }
  }

  // Check for invalid maxEntities
  if (config.filters?.maxEntities !== undefined && config.filters.maxEntities < 1) {
    errors.push('maxEntities must be greater than 0');
  }

  return errors;
}

/**
 * Apply populate configuration to entities in the graph
 */
export function applyPopulateConfig(
  graph: JsonLdGraph,
  populateConfig: PopulateConfig
): JsonLdGraph {
  return graph.map((entity) => {
    const entityId = entity['@id'];
    const populateRules = populateConfig[entityId];

    if (!populateRules || populateRules.length === 0) {
      return entity;
    }

    // Create a copy of the entity to avoid mutation
    const populatedEntity = { ...entity };

    // Apply each populate rule
    for (const rule of populateRules) {
      populatedEntity[rule.property] = rule.entities;
    }

    return populatedEntity;
  });
}

/**
 * Validate runtime configuration for critical errors only
 * This is more lenient than the general validateConfig to handle runtime overrides
 */
export function validateRuntimeConfig(config: JsonLdConfig): string[] {
  const errors: string[] = [];

  // Check for baseGraph requirement
  if (!config.baseGraph) {
    errors.push('baseGraph is required');
  } else if (!Array.isArray(config.baseGraph)) {
    errors.push('baseGraph must be an array');
  }

  // Only check for critical errors that would break processing
  if (config.filters?.maxEntities !== undefined && config.filters.maxEntities < 1) {
    errors.push('maxEntities must be greater than 0');
  }

  return errors;
}

/**
 * Type guard to check if an entity matches a type
 */
export function entityHasType(entity: JsonLdEntity, type: string): boolean {
  const entityType = entity['@type'];
  if (!entityType) return false;

  if (Array.isArray(entityType)) {
    return entityType.includes(type);
  }

  return entityType === type;
}

/**
 * Type guard to check if an entity has any of the specified types
 */
export function entityHasAnyType(entity: JsonLdEntity, types: string[]): boolean {
  return types.some((type) => entityHasType(entity, type));
}

/**
 * Get all types from an entity as an array
 */
export function getEntityTypes(entity: JsonLdEntity): string[] {
  const entityType = entity['@type'];
  if (!entityType) return [];

  return Array.isArray(entityType) ? entityType : [entityType];
}

/**
 * Apply filters to a JSON-LD graph (copied from head-manager/filters.ts for self-containment)
 */
export function filterJsonLdGraph(graph: JsonLdGraph, options: JsonLdFilterOptions): JsonLdGraph {
  let filtered = [...graph];

  // Apply type inclusion filter
  if (options.includeTypes && options.includeTypes.length > 0) {
    filtered = filtered.filter((entity) => {
      const type = entity['@type'];
      if (!type) return false;

      const types = Array.isArray(type) ? type : [type];
      return types.some((t) => options.includeTypes!.includes(t));
    });
  }

  // Apply type exclusion filter
  if (options.excludeTypes && options.excludeTypes.length > 0) {
    filtered = filtered.filter((entity) => {
      const type = entity['@type'];
      if (!type) return true;

      const types = Array.isArray(type) ? type : [type];
      return !types.some((t) => options.excludeTypes!.includes(t));
    });
  }

  // Apply ID inclusion filter
  if (options.includeIds && options.includeIds.length > 0) {
    filtered = filtered.filter(
      (entity) => entity['@id'] && options.includeIds!.includes(entity['@id'])
    );
  }

  // Apply ID exclusion filter
  if (options.excludeIds && options.excludeIds.length > 0) {
    filtered = filtered.filter(
      (entity) => !entity['@id'] || !options.excludeIds!.includes(entity['@id'])
    );
  }

  // Apply required properties filter
  if (options.requiredProperties && options.requiredProperties.length > 0) {
    filtered = filtered.filter((entity) =>
      options.requiredProperties!.every((prop) => prop in entity)
    );
  }

  // Apply excluded properties filter (handle both old and new property names)
  const excludeProps = options.excludeEntitiesWithProperties || (options as any).excludeProperties;
  if (excludeProps && excludeProps.length > 0) {
    filtered = filtered.filter((entity) => !excludeProps.some((prop: string) => prop in entity));
  }

  // Apply custom filter
  if (options.customFilter) {
    filtered = filtered.filter(options.customFilter);
  }

  // Apply max entities limit
  if (options.maxEntities && options.maxEntities > 0) {
    filtered = filtered.slice(0, options.maxEntities);
  }

  return filtered;
}
