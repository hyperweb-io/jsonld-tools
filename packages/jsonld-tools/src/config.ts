/**
 * Immutable configuration builder for JSON-LD processing
 *
 * Provides a fluent interface for building configurations without
 * performing any graph processing. Each method returns a new
 * configuration object for safe reuse and extension.
 */

import type {
	IJsonLdConfigBuilder,
	JsonLdConfig,
	JsonLdEntity,
	JsonLdFilter,
	JsonLdGraph,
	PipeFunction,
	PopulateConfig,
	PropertyFilterByIdRule,
	PropertyFilterByTypeRule,
} from './types';

/**
 * Immutable configuration builder class
 */
export class JsonLdConfigBuilder implements IJsonLdConfigBuilder {
	private config: JsonLdConfig;

	constructor(initialConfig: JsonLdConfig = {}) {
		this.config = { ...initialConfig };
	}

	/**
	 * Create a new instance of this class (can be overridden by subclasses)
	 */
	protected createInstance(config: JsonLdConfig): JsonLdConfigBuilder {
		return new JsonLdConfigBuilder(config);
	}

	/**
	 * Set the base graph to process
	 */
	baseGraph(graph: JsonLdGraph): this {
		return this.createInstance({
			...this.config,
			baseGraph: graph,
		}) as this;
	}

	/**
	 * Include only entities with these IDs (merges with existing IDs)
	 */
	includeIds(ids: string[]): this {
		const existingIds = this.config.filters?.includeIds || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				includeIds: [...existingIds, ...ids],
			},
		}) as this;
	}

	/**
	 * Exclude entities with these IDs (merges with existing IDs)
	 */
	excludeIds(ids: string[]): this {
		const existingIds = this.config.filters?.excludeIds || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				excludeIds: [...existingIds, ...ids],
			},
		}) as this;
	}

	/**
	 * Include only these entity types (merges with existing types)
	 */
	includeTypes(types: string[]): this {
		const existingTypes = this.config.filters?.includeTypes || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				includeTypes: [...existingTypes, ...types],
			},
		}) as this;
	}

	/**
	 * Exclude these entity types (merges with existing types)
	 */
	excludeTypes(types: string[]): this {
		const existingTypes = this.config.filters?.excludeTypes || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				excludeTypes: [...existingTypes, ...types],
			},
		}) as this;
	}

	/**
	 * Apply custom filter function
	 */
	customFilter(fn: JsonLdFilter): this {
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				customFilter: fn,
			},
		}) as this;
	}

	/**
	 * Limit maximum number of entities
	 */
	maxEntities(max: number): this {
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				maxEntities: max,
			},
		}) as this;
	}

	/**
	 * Include only entities with these properties (merges with existing properties)
	 * @param props - Array of property names that entities must have
	 * @example
	 * ```typescript
	 * config.requiredProperties(['name']).requiredProperties(['url'])
	 * // Result: entities must have both 'name' and 'url' properties
	 * ```
	 */
	requiredProperties(props: string[]): this {
		const existingProps = this.config.filters?.requiredProperties || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				requiredProperties: [...existingProps, ...props],
			},
		}) as this;
	}

	/**
	 * Exclude entities with these properties (merges with existing properties)
	 */
	excludeEntitiesWithProperties(props: string[]): this {
		const existingProps = this.config.filters?.excludeEntitiesWithProperties || [];
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				excludeEntitiesWithProperties: [...existingProps, ...props],
			},
		}) as this;
	}

	/**
	 * Extract subgraph starting from these root IDs
	 */
	subgraph(rootIds: string[]): this {
		return this.createInstance({
			...this.config,
			subgraphRoots: [...(this.config?.subgraphRoots || []), ...rootIds],
		}) as this;
	}

	/**
	 * Filter properties for specific entity IDs
	 */
	filterPropertiesByIds(entityIds: string[], rule: { include?: string[]; exclude?: string[] }): this {
		const existingRules = this.config.propertyFiltersByIds || [];
		const newRule: PropertyFilterByIdRule = {
			entityIds,
			...rule,
		};

		return this.createInstance({
			...this.config,
			propertyFiltersByIds: [...existingRules, newRule],
		}) as this;
	}

	/**
	 * Filter properties for specific entity types
	 */
	filterPropertiesByTypes(entityTypes: string[], rule: { include?: string[]; exclude?: string[] }): this {
		const existingRules = this.config.propertyFiltersByTypes || [];
		const newRule: PropertyFilterByTypeRule = {
			entityTypes,
			...rule,
		};

		return this.createInstance({
			...this.config,
			propertyFiltersByTypes: [...existingRules, newRule],
		}) as this;
	}

	/**
	 * Add additional entities
	 */
	addEntities(entities: JsonLdEntity[]): this {
		const existingEntities = this.config.additionalEntities || [];
		return this.createInstance({
			...this.config,
			additionalEntities: [...existingEntities, ...entities],
		}) as this;
	}

	/**
	 * Add custom transformation pipe
	 */
	pipe(fn: PipeFunction): this {
		const existingPipes = this.config.pipes || [];
		return this.createInstance({
			...this.config,
			pipes: [...existingPipes, fn],
		}) as this;
	}

	/**
	 * Configure entity population
	 */
	populateEntities(entityIds: string[], populateConfig: PopulateConfig): this {
		// Merge with existing populate config
		const existingPopulateConfig = this.config.populateConfig || {};
		const mergedPopulateConfig = { ...existingPopulateConfig };

		// Add the new populate config for the specified entity IDs
		for (const entityId of entityIds) {
			if (populateConfig[entityId]) {
				mergedPopulateConfig[entityId] = populateConfig[entityId];
			}
		}

		return this.createInstance({
			...this.config,
			populateConfig: mergedPopulateConfig,
		}) as this;
	}

	/**
	 * Clear both includeIds and excludeIds
	 * @example
	 * ```typescript
	 * config.includeIds(['a']).excludeIds(['b']).clearIds().includeIds(['c'])
	 * // Result: only includes 'c', previous IDs are cleared
	 * ```
	 */
	clearIds(): this {
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				includeIds: undefined,
				excludeIds: undefined,
			},
		}) as this;
	}

	/**
	 * Clear both includeTypes and excludeTypes
	 */
	clearTypes(): this {
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				includeTypes: undefined,
				excludeTypes: undefined,
			},
		}) as this;
	}

	/**
	 * Clear both requiredProperties and excludeEntitiesWithProperties
	 */
	clearPropertyRequirements(): this {
		return this.createInstance({
			...this.config,
			filters: {
				...this.config.filters,
				requiredProperties: undefined,
				excludeEntitiesWithProperties: undefined,
			},
		}) as this;
	}

	/**
	 * Clear both propertyFiltersByIds and propertyFiltersByTypes
	 */
	clearPropertyFilters(): this {
		return this.createInstance({
			...this.config,
			propertyFiltersByIds: undefined,
			propertyFiltersByTypes: undefined,
		}) as this;
	}

	/**
	 * Clear subgraphRoots
	 */
	clearSubgraph(): this {
		return this.createInstance({
			...this.config,
			subgraphRoots: undefined,
		}) as this;
	}

	/**
	 * Clear entire configuration except baseGraph
	 * @example
	 * ```typescript
	 * config.baseGraph(graph).includeIds(['a']).clearAll().includeTypes(['Person'])
	 * // Result: only baseGraph and includeTypes remain
	 * ```
	 */
	clearAll(): this {
		return this.createInstance({
			baseGraph: this.config.baseGraph,
		}) as this;
	}

	/**
	 * Get the current configuration
	 */
	getConfig(): JsonLdConfig {
		return { ...this.config };
	}
}

/**
 * Factory function to create a new configuration builder
 */
export function createJsonLdConfig(): JsonLdConfigBuilder {
	return new JsonLdConfigBuilder();
}
