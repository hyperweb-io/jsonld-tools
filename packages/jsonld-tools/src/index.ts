export { createJsonLdConfig, JsonLdConfigBuilder } from './config';
export { createJsonLdBuilder, JsonLdBuilder } from './builder';

export type { JsonLdConfig, JsonLdFilterOptions, BuildOptions, PopulateConfig } from './types';

export {
  type JsonLdEntity,
  type JsonLdGraph,
  extractReferences,
  extractSubgraph,
  extractSubgraphWithDepth,
  extractSubgraphs,
  filterEntityProperties,
  filterGraphProperties,
  findEntities,
  findEntitiesByType,
  findEntity,
  findMissingReferences,
  findNestedEntities,
  findOrphans,
  findReferencingEntities,
  inlineReferences,
} from './jsonld-utils';

export { filterJsonLdGraph } from './builder-utils';
