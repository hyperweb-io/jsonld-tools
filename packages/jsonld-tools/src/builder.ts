/**
 * JSON-LD Builder with fluent interface
 *
 * Provides a mutable builder that applies configurations and processes graphs.
 * Supports runtime overrides and maintains the complex filtering logic from
 * the current head-manager system.
 */

import { JsonLdConfigBuilder } from './config';
import { extractSubgraphs, filterGraphProperties, type JsonLdGraph } from './jsonld-utils';
import type { BuildOptions, JsonLdConfig } from './types';
import {
  applyPopulateConfig,
  buildPropertyFilterConfig,
  filterJsonLdGraph,
  mergeConfigs,
  validateRuntimeConfig,
} from './builder-utils';

/**
 * Mutable builder class that extends config builder with automatic method delegation
 */
export class JsonLdBuilder extends JsonLdConfigBuilder {
  private currentGraph?: JsonLdGraph;

  constructor(initialConfig: JsonLdConfig = {}) {
    super(initialConfig);
  }

  /**
   * Override the config builder's method creation to return JsonLdBuilder instances
   */
  protected createInstance(config: JsonLdConfig): JsonLdBuilder {
    return new JsonLdBuilder(config);
  }

  /**
   * Apply a pre-built configuration (builder-specific method)
   */
  applyConfig(config: JsonLdConfig): JsonLdBuilder {
    const currentConfig = this.getConfig();
    const mergedConfig = mergeConfigs(currentConfig, config);
    return new JsonLdBuilder(mergedConfig);
  }

  /**
   * Get the current graph state (processes the graph if not already done)
   */
  getCurrentGraph(): JsonLdGraph {
    if (!this.currentGraph) {
      this.currentGraph = this.processGraph();
    }
    return this.currentGraph;
  }

  /**
   * Build the final JSON-LD output
   * @param options - Build configuration options
   * @param options.prettyPrint - Whether to format JSON with indentation (default: true)
   * @param options.contextUrl - The JSON-LD context URL (default: 'https://schema.org')
   * @param options.withScriptTag - Whether to wrap output in a script tag (default: false)
   * @param options.scriptId - Optional ID attribute for the script tag
   * @returns The JSON-LD string, optionally wrapped in a script tag
   */
  build(options: BuildOptions = {}): string {
    const {
      prettyPrint = true,
      contextUrl = 'https://schema.org',
      withScriptTag = false,
      scriptId,
    } = options;

    const graph = this.getCurrentGraph();

    // Create the JSON-LD object
    const jsonLd = {
      '@context': contextUrl,
      '@graph': graph,
    };

    // Stringify with optional pretty printing
    const jsonString = JSON.stringify(jsonLd, null, prettyPrint ? 2 : undefined);

    // Wrap in script tag if requested
    if (withScriptTag) {
      const idAttribute = scriptId ? ` id="${scriptId}"` : '';
      return `<script${idAttribute} type="application/ld+json">${jsonString}</script>`;
    }

    return jsonString;
  }

  /**
   * Process the graph according to the current configuration
   * This implements the critical filtering logic from the current manager
   */
  private processGraph(): JsonLdGraph {
    const config = this.getConfig();

    if (!config.baseGraph) {
      throw new Error('No base graph provided. Use baseGraph() or applyConfig() with a baseGraph.');
    }

    // Validate configuration - only check for critical errors that would break processing
    const errors = validateRuntimeConfig(config);
    if (errors.length > 0) {
      throw new Error(`Configuration validation failed: ${errors.join(', ')}`);
    }

    let graph = [...config.baseGraph];
    const propertyFilters = buildPropertyFilterConfig(config);
    let propertyFiltersApplied = false;

    // Layer 1: Subgraph extraction with property filtering if subgraphRoots are set
    if (config.subgraphRoots && config.subgraphRoots.length > 0) {
      // Extract subgraph with property filtering during traversal
      graph = extractSubgraphs(graph, config.subgraphRoots, propertyFilters);
      propertyFiltersApplied = true; // Property filters were applied during subgraph extraction
    }

    // Layer 2: Regular entity filtering with property filtering if not already applied
    if (config.filters) {
      // Apply property filters first if they haven't been applied yet
      if (!propertyFiltersApplied && propertyFilters) {
        graph = filterGraphProperties(graph, propertyFilters);
        propertyFiltersApplied = true;
      }

      // Apply entity-level filters
      graph = filterJsonLdGraph(graph, config.filters);
    }

    // Apply entity population if configured
    if (config.populateConfig) {
      graph = applyPopulateConfig(graph, config.populateConfig);
    }

    // Add additional entities if specified
    if (config.additionalEntities && config.additionalEntities.length > 0) {
      graph = [...graph, ...config.additionalEntities];
    }

    // Apply custom transformation pipes
    if (config.pipes && config.pipes.length > 0) {
      for (const pipe of config.pipes) {
        graph = pipe(graph);
      }
    }

    return graph;
  }
}

/**
 * Factory function to create a new builder
 */
export function createJsonLdBuilder(): JsonLdBuilder {
  return new JsonLdBuilder();
}
