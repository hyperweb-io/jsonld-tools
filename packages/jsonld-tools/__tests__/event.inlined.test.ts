import { extractSubgraph, inlineReferences } from '../src/index';
import graph from '../../../fixtures/graph.json';

it('event:osmocon-2023-paris expanded with inlined references', async () => {
  // Extract the subgraph containing the event and all its references
  const subgraph = extractSubgraph(graph, 'event:osmocon-2023-paris');
  
  // Manually inline all @id references with their full entity data
  const inlined = inlineReferences(subgraph, 'event:osmocon-2023-paris');
  
  expect(inlined).toMatchSnapshot();
});
