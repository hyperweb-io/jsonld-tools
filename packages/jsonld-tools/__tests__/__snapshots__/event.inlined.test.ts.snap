// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event:osmocon-2023-paris expanded with inlined references 1`] = `
{
  "@id": "event:osmocon-2023-paris",
  "@type": "Event",
  "description": "Annual Osmosis conference in Paris",
  "endDate": "2023-10-17",
  "location": "Paris, France",
  "name": "Osmocon 2023 Paris",
  "organizer": {
    "@id": "org:osmosis",
    "@type": "Organization",
    "foundingDate": "2021-06-01",
    "member": [
      {
        "@id": "person:john",
        "@type": [
          "Person",
          "Developer",
        ],
        "email": "john@example.com",
        "jobTitle": "Senior Developer",
        "memberOf": [
          {
            "@id": "org:hyperweb",
            "@type": "Organization",
            "foundingDate": "2020-01-01",
            "location": {
              "@id": "place:san-francisco",
              "@type": "Place",
              "address": {
                "@type": "PostalAddress",
                "addressCountry": "US",
                "addressLocality": "San Francisco",
                "addressRegion": "CA",
              },
              "geo": {
                "@type": "GeoCoordinates",
                "latitude": 37.7749,
                "longitude": -122.4194,
              },
              "name": "San Francisco",
            },
            "member": [
              {
                "@id": "person:danlynch",
                "@type": "Person",
                "affiliation": {
                  "@id": "org:hyperweb",
                },
                "email": "dan@hyperweb.com",
                "jobTitle": "Founder & CEO",
                "knows": [
                  "person:jane",
                  "person:john",
                ],
                "name": "Dan Lynch",
                "worksFor": "org:hyperweb",
              },
              {
                "@id": "person:john",
              },
            ],
            "name": "HyperWeb",
            "url": "https://hyperweb.com",
          },
          {
            "@id": "org:osmosis",
          },
        ],
        "name": "John Smith",
      },
    ],
    "name": "Osmosis",
    "url": "https://osmosis.zone",
  },
  "performer": [
    "person:danlynch",
    "person:jane",
  ],
  "startDate": "2023-10-15",
}
`;
